### PR TITLE
ENH: support NaT values into datetime series for interpolation (#11701)

### LINKED ISSUE #11701: Interpolate NaT
Hello,

`interpolate` doesn't work with `NaT`
see http://stackoverflow.com/questions/33921795/fill-timestamp-nat-with-a-linear-interpolation/33922824#33922824

Here is a trivial example to show the situation:

```
s = pd.Series(pd.date_range('2015-01-01' , '2015-01-30'), name='t')

s[3], s[4], s[5] = pd.NaT, pd.NaT, pd.NaT
s[13], s[14], s[15] = pd.NaT, pd.NaT, pd.NaT
print(s)

0    2015-01-01
1    2015-01-02
2    2015-01-03
3           NaT
4           NaT
5           NaT
6    2015-01-07
7    2015-01-08
8    2015-01-09
9    2015-01-10
10   2015-01-11
11   2015-01-12
12   2015-01-13
13          NaT
14          NaT
15          NaT
16   2015-01-17
17   2015-01-18
18   2015-01-19
19   2015-01-20
20   2015-01-21
21   2015-01-22
22   2015-01-23
23   2015-01-24
24   2015-01-25
25   2015-01-26
26   2015-01-27
27   2015-01-28
28   2015-01-29
29   2015-01-30
Name: t, dtype: datetime64[ns]

print(s.interpolate())
0    2015-01-01
1    2015-01-02
2    2015-01-03
3           NaT
4           NaT
5           NaT
6    2015-01-07
7    2015-01-08
8    2015-01-09
9    2015-01-10
10   2015-01-11
11   2015-01-12
12   2015-01-13
13          NaT
14          NaT
15          NaT
16   2015-01-17
17   2015-01-18
18   2015-01-19
19   2015-01-20
20   2015-01-21
21   2015-01-22
22   2015-01-23
23   2015-01-24
24   2015-01-25
25   2015-01-26
26   2015-01-27
27   2015-01-28
28   2015-01-29
29   2015-01-30
Name: t, dtype: datetime64[ns]

assert s.interpolate().isnull().sum() == 0
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-150-8a59e397a174> in <module>()
----> 1 assert s.interpolate().isnull().sum() == 0

AssertionError:
```

Kind regards

### PR DESCRIPTION
**Pull request by atulagrwl**
- [ ] closes #11701
- [ ] tests added / passed
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

**Comment by jbrockmendel** [comment]
Glad this is getting fixed, good work.  Want to handle the time delta case while you’re at it?

**Comment by atulaggarwal** [comment]
@jbrockmendel - Are you referring https://github.com/pandas-dev/pandas/issues/19199 ?

**Comment by jbrockmendel** [comment]
#19199 is also about datetime64.  I'm not sure if there is a separate issue for `timedelta64[ns]`, but the behavior is the same.

**Comment by atulaggarwal** [comment]
Okay. I will take a look at both #19199 and timedelta64[ns]

File: pandas/tests/series/test_missing.py [hunk 1]
```diff
@@ -1317,3 +1317,23 @@ def test_series_interpolate_intraday(self):
         result = ts.reindex(new_index).interpolate(method='time')
 
         tm.assert_numpy_array_equal(result.values, exp.values)
+
+    def test_series_interpolate_nat(self):
```
Comment (jreback) [inline pandas/tests/series/test_missing.py]: use tz_native_fixture and paramerize

File: pandas/tests/series/test_missing.py [hunk 2]
```diff
@@ -1317,3 +1317,23 @@ def test_series_interpolate_intraday(self):
...
+        # GH 11701
+        for tz in [None, 'UTC', 'Europe/Paris']:
+            expected = pd.Series(pd.date_range('2015-01-01',
+                                               '2015-01-30', tz=tz))
+            result = expected.copy()
+            result[[3, 4, 5, 13, 14, 15]] = pd.NaT
+            result = result.interpolate()
+            tm.assert_series_equal(result, expected)
+
+    def test_series_interpolate_nat_inplace(self):
```
Comment (jreback) [inline pandas/tests/series/test_missing.py]: same

Comment (jreback) [reply pandas/tests/series/test_missing.py]: you can parameterize over inplace as well

File: pandas/core/generic.py [hunk 3]
```diff
@@ -6112,6 +6112,14 @@ def interpolate(self, method='linear', axis=0, limit=None, inplace=False,
             raise NotImplementedError("Interpolation with NaNs in the index "
                                       "has not been implemented. Try filling "
                                       "those NaNs before interpolating.")
+        is_datetime = False
```
Comment (jreback) [inline pandas/core/generic.py]: this needs to be pushed down to the block in ``pandas/core/internals``, should go with ``def _interpolate`` I think

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/series/test_missing.py (hunk 1 above)
Comment (atulaggarwal) [reply pandas/tests/series/test_missing.py]: Fixed in latest commit.

File: pandas/core/generic.py (hunk 3 above)
Comment (atulaggarwal) [reply pandas/core/generic.py]: I have to explore more about it. Will do that in a day or two.

**Comment by jbrockmendel** [comment]
@atulaggarwal needs rebase.  We can try to push this through as is, or aim for the cleaner long-term solution of implementing interpolate on DatetimeArray and TimedeltaArray.

**Comment by atulaggarwal** [comment]
@jbrockmendel - Updated PR with latest master. I would suggest merging it as is. I will work on working on other items too (but lets not block on them) as current PR solves one of the problem. What do you suggest?

File: doc/source/whatsnew/v0.24.0.txt [hunk 4]
```diff
@@ -182,6 +182,8 @@ Other Enhancements
 - :func:`to_timedelta` now supports iso-formated timedelta strings (:issue:`21877`)
 - :class:`Series` and :class:`DataFrame` now support :class:`Iterable` in constructor (:issue:`2193`)
 - :class:`DatetimeIndex` gained :attr:`DatetimeIndex.timetz` attribute. Returns local time with timezone information. (:issue:`21358`)
+- Implement interpolating ``NaT`` values in ``datetime`` series (:issue:`11701`)
```
Comment (jbrockmendel) [inline doc/source/whatsnew/v0.24.0.txt]: The wording here is awkward.  It doesn't need to be a complete sentence per se, but take a look at the other entries for examples.

**Comment by jbrockmendel** [comment]
> I would suggest merging it as is.

As jreback mentioned, the implementation is not in the right place.  The viable options are in `core.internals.blocks.DatetimeLikeBlockMixin` and `core.arrays.datetimelike.DatetimeLikeArrayMixin`

**Comment by jbrockmendel** [comment]
I think this might give incorrect answers for values near the int64 bounds because casting to float will be lossy.  Can you add a test for that?

**Comment by jreback** [comment]
closing as stale. if you'd like to continue, pls ping.

**Pull request closed without merging**